src/libraries/System.Private.CoreLib/src/System/Array.cs

```diff
@@ -949,7 +949,7 @@ public static int BinarySearch(Array array, int index, int length, object? value
                         return (result >= 0) ? (index + result) : ~(index + ~result);
 
                         static int GenericBinarySearch<T>(Array array, int adjustedIndex, int length, object value) where T : struct, IComparable<T>
-                            => UnsafeArrayAsSpan<T>(array, adjustedIndex, length).BinarySearch(Unsafe.As<byte, T>(ref value.GetRawData()));
+                            => UnsafeArrayAsSpan<T>(array, adjustedIndex, length).BinarySearch(Unsafe.ReadUnaligned<T>(ref value.GetRawData()));
```

Comment:
aren't array's elements always aligned to T here?

Reply:
doesn't hurt to be obviously correct, also it's 5 bytes less IL

Reply:
I wasn't sure about the data alignment of e.g. `double` on all 32 bit platforms.

Reply:
> doesn't hurt to be obviously correct, also it's 5 bytes less IL

The problem of using ReadUnaligned where it's not needed is a theoretical serious performance regression on platforms where ReadUnaligned is implemented using memcpy-like routine (due to slow/faulty unaligned access)

Reply:
There is no alignment problem here. 

> doesn't hurt to be obviously correct

I think it is confusing to use ReadUnaligned in place when there are no alignment issues.

> I guess since https://github.com/dotnet/runtime/pull/115985 it's not guaranteed

`double`s are not 8-byte aligned on x86 and you do not need to worry about it. Regular double loads cannot use instructions that except 8-byte alignment on x86.

Reply:
Could you please revert these unnecessary ReadUnaligned? The BitCast changes are good.